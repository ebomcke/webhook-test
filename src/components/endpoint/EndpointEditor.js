import React, { Component } from 'react';
import { compose } from 'recompose';
import { withFirestore } from 'react-firestore';
import { withFirebase } from '../../contexts/Firebase';
import { withAccount } from '../../contexts/Account';
import { withNavigation } from '../../contexts/Navigation';

class EndpointEditor extends Component {
  constructor(props) {
    super(props);
    let { endpoint } = this.props;
    if (!endpoint) {
      endpoint = this.generateNewEndpoint();
    }
    this.state = {
      id: endpoint.id,
      path: endpoint.path,
      defaultResponseBody: endpoint.defaultResponse.body,
      defaultResponseStatusCode: endpoint.defaultResponse.statusCode,
    };
  }

  generateNewEndpoint = () => ({
    path: '',
    defaultResponse: {
      statusCode: 200,
      body: JSON.stringify(
        {
          status: 'OK',
        },
        null,
        2,
      ),
    },
  });

  onStatusCodeChange = (statusCode) => {
    this.setState({
      defaultResponseStatusCode: statusCode,
    });
  }

  onUrlChange = (url) => {
    url = url
      .toLowerCase()
      .replace(/ /gi, '')
      .replace(/^\//, '');
    this.setState({
      path: url,
    });
  }

  onResponseBodyChange = (value) => {
    try {
      JSON.parse(value);
      this.setState({
        defaultResponseBody: value,
      });
    } catch (e) {}
  }

  save = async () => {
    const { firebase, firestore, account, navigate } = this.props;
    let endpoint = {
      account: firestore.collection('accounts').doc(account.id),
      path: this.state.path,
      lastActive: firebase.firestore.Timestamp.now(),
      defaultResponse: {
        statusCode: this.state.defaultResponseStatusCode,
        body: this.state.defaultResponseBody,
      },
    };
    if (this.state.id) {
      endpoint.id = this.state.id;
    }
    if (endpoint.id) {
      await firestore
        .collection('endpoints')
        .doc(endpoint.id)
        .set(endpoint);
    } else {
      await firestore.collection('endpoints').add(endpoint);
    }
    navigate({
      name: 'Endpoints',
    });
  }

  render() {
    const { children } = this.props;
    return children({
      endpoint: this.state,
      onResponseBodyChange: this.onResponseBodyChange,
      onUrlChange: this.onUrlChange,
      onStatusCodeChange: this.onStatusCodeChange,
      save: this.save,
    });
  }
}

export default compose(
  withFirebase,
  withFirestore,
  withNavigation,
  withAccount,
)(EndpointEditor);
