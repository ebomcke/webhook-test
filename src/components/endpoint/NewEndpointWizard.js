import React, { Component } from 'react';
import { Segment, Header, Icon, Message } from 'semantic-ui-react';
import { compose } from 'recompose';
import { withNavigation } from '../../contexts/Navigation';
import { withAccount } from '../../contexts/Account';
import Wizard from '../base/Wizard';
import { EditEndpointUrl, EditEndpointResponse } from './EditEndpoint';

class NewEndpointWizard extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    const { endpointEditor } = this.props;
    const {
      endpoint,
    } = endpointEditor;
    const steps = [
      {
        key: 'url',
        icon: 'at',
        title: 'URL',
        description: 'Choose a URL',
        render: nextButton => (
          <EditEndpointUrl
            endpointEditor={endpointEditor}
            nextButton={nextButton}
          />
        ),
        nextDisabled: () => !endpoint.path,
      },
      {
        key: 'response',
        icon: 'code',
        title: 'Response',
        description: 'Set the default response',
        render: nextButton => (
          <EditEndpointResponse renderMessage={true}
            endpointEditor={endpointEditor}
            nextButton={nextButton}
          />
        ),
      },
      {
        key: 'done',
        title: "You're done",
        render: nextButton => (
          <div>
            <Message icon positive>
              <Icon name="checkmark" />
              <Message.Content>
                <Message.Header>Congratulations !</Message.Header>
                <p>You are now ready to receive your first webhook.</p>
              </Message.Content>
            </Message>
            {nextButton()}
          </div>
        ),
      },
    ];
    return (
      <Segment className="full-height">
        <Header as="h1">New endpoint</Header>
        <Wizard steps={steps} onComplete={endpointEditor.save} />
      </Segment>
    );
  }
}

export default compose(
  withNavigation,
  withAccount,
)(NewEndpointWizard);
