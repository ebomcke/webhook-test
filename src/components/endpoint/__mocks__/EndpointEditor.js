const mockEndpointEditor = {
    endpoint: {
        path: 'unit-test-path',
        defaultResponse: {
            statusCode: 200,
            body: JSON.stringify({status:'OK'})
        }
    },
    onResponseBodyChange: jest.fn(),
    onUrlChange: jest.fn(),
    onStatusCodeChange: jest.fn(),
    save: jest.fn(() => Promise.resolve())
}

export { mockEndpointEditor };