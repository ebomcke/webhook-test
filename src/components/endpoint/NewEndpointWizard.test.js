import React from 'react';
import { create } from 'react-test-renderer';
import NewEndpointWizard from './NewEndpointWizard';

jest.mock('./EndpointEditor');
import { mockEndpointEditor } from './EndpointEditor';

describe('Rendering tests', () => {
  test('Step 1 - endpoint url', () => {
    const newEndpointWizard = create(
      <NewEndpointWizard endpointEditor={mockEndpointEditor} />,
    );
    expect(newEndpointWizard.toJSON()).toMatchSnapshot();
  });

  test('Step 2 - default response', () => {
    const newEndpointWizard = create(
      <NewEndpointWizard endpointEditor={mockEndpointEditor} />,
    );
    newEndpointWizard.root
      .findByProps({
        wizardIdentifier: 'next-done-button',
      })
      .props.onClick();
    expect(newEndpointWizard.toJSON()).toMatchSnapshot();
  });

  test('Step 3 - done', () => {
    const newEndpointWizard = create(
      <NewEndpointWizard endpointEditor={mockEndpointEditor} />,
    );
    newEndpointWizard.root
      .findByProps({
        wizardIdentifier: 'next-done-button',
      })
      .props.onClick();
    newEndpointWizard.root
      .findByProps({
        wizardIdentifier: 'next-done-button',
      })
      .props.onClick();
    expect(newEndpointWizard.toJSON()).toMatchSnapshot();
  });
});

describe('Interactions tests', () => {
  afterEach(() => {
    mockEndpointEditor.save.mockClear();
  });
  test('Save triggers endpoint save', async () => {
    const newEndpointWizard = create(
      <NewEndpointWizard endpointEditor={mockEndpointEditor} />,
    );
    await newEndpointWizard.root
      .findByProps({
        wizardIdentifier: 'next-done-button',
      })
      .props.onClick();
    await newEndpointWizard.root
      .findByProps({
        wizardIdentifier: 'next-done-button',
      })
      .props.onClick();
    await newEndpointWizard.root
      .findByProps({
        wizardIdentifier: 'next-done-button',
      })
      .props.onClick();
    expect(mockEndpointEditor.save.mock.calls.length).toBe(1);
  });
});
