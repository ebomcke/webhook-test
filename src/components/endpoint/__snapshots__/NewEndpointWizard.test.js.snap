// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering tests Step 1 - endpoint url 1`] = `
<div
  className="ui segment full-height"
>
  <h1
    className="ui header"
  >
    New endpoint
  </h1>
  <div>
    <div
      className="ui fluid steps"
    >
      <div
        className="active step"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="at icon"
        />
        <div
          className="content"
        >
          <div
            className="title"
          >
            URL
          </div>
          <div
            className="description"
          >
            Choose a URL
          </div>
        </div>
      </div>
      <div
        className="disabled step"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="code icon"
        />
        <div
          className="content"
        >
          <div
            className="title"
          >
            Response
          </div>
          <div
            className="description"
          >
            Set the default response
          </div>
        </div>
      </div>
      <div
        className="disabled step"
        onClick={[Function]}
      >
        <div
          className="content"
        >
          <div
            className="title"
          >
            You're done
          </div>
        </div>
      </div>
    </div>
    <form
      className="ui form"
      onSubmit={[Function]}
    >
      <div
        className="required field"
      >
        <label>
          URL
        </label>
        <div
          className="ui labeled input"
        >
          <div
            className="ui label label"
            onClick={[Function]}
          >
            https://webhook-tester-1547141127436.appspot.com/webhook/unit-test-org/
          </div>
          <input
            name="path"
            onChange={[Function]}
            placeholder="my/custom/url"
            type="text"
            value="unit-test-path"
          />
        </div>
      </div>
      <div
        className="field"
      >
        <button
          className="ui button"
          onClick={[Function]}
          wizardIdentifier="next-done-button"
        >
          Next
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Rendering tests Step 2 - default response 1`] = `
<div
  className="ui segment full-height"
>
  <h1
    className="ui header"
  >
    New endpoint
  </h1>
  <div>
    <div
      className="ui fluid steps"
    >
      <div
        className="step"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="at icon"
        />
        <div
          className="content"
        >
          <div
            className="title"
          >
            URL
          </div>
          <div
            className="description"
          >
            Choose a URL
          </div>
        </div>
      </div>
      <div
        className="active step"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="code icon"
        />
        <div
          className="content"
        >
          <div
            className="title"
          >
            Response
          </div>
          <div
            className="description"
          >
            Set the default response
          </div>
        </div>
      </div>
      <div
        className="disabled step"
        onClick={[Function]}
      >
        <div
          className="content"
        >
          <div
            className="title"
          >
            You're done
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="ui icon message"
      >
        <i
          aria-hidden="true"
          className="help icon"
        />
        <div
          className="content"
        >
          <div
            className="header"
          >
            This is optional
          </div>
          <p>
            If you're not sure what the default response should be you can safely leave this default value.
          </p>
        </div>
      </div>
      <form
        className="ui form"
        onSubmit={[Function]}
      >
        <div
          className="field"
        >
          <div
            className="field"
          >
            <label>
              Status code
            </label>
            <div
              aria-expanded={false}
              className="ui fluid selection dropdown"
              name="defaultResponseStatusCode"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              role="listbox"
              tabIndex={0}
            >
              <div
                aria-live="polite"
                className="text"
                role="alert"
              />
              <i
                aria-hidden="true"
                className="dropdown icon"
                onClick={[Function]}
              />
              <div
                className="menu transition"
              >
                <div
                  aria-checked={false}
                  aria-selected={true}
                  className="selected item"
                  onClick={[Function]}
                  role="option"
                  style={
                    Object {
                      "pointerEvents": "all",
                    }
                  }
                >
                  <span
                    className="text"
                  >
                    200 - OK
                  </span>
                </div>
                <div
                  aria-checked={false}
                  aria-selected={false}
                  className="item"
                  onClick={[Function]}
                  role="option"
                  style={
                    Object {
                      "pointerEvents": "all",
                    }
                  }
                >
                  <span
                    className="text"
                  >
                    201 - Created
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="field"
        >
          <label>
            Response body
          </label>
          <div>
            This is a placeholder for the JSON editor.
            <pre>
              <code>
                {"height":"300px"}
              </code>
            </pre>
          </div>
        </div>
        <div
          className="field"
        >
          <button
            className="ui button"
            onClick={[Function]}
            wizardIdentifier="next-done-button"
          >
            Next
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`Rendering tests Step 3 - done 1`] = `
<div
  className="ui segment full-height"
>
  <h1
    className="ui header"
  >
    New endpoint
  </h1>
  <div>
    <div
      className="ui fluid steps"
    >
      <div
        className="step"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="at icon"
        />
        <div
          className="content"
        >
          <div
            className="title"
          >
            URL
          </div>
          <div
            className="description"
          >
            Choose a URL
          </div>
        </div>
      </div>
      <div
        className="step"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="code icon"
        />
        <div
          className="content"
        >
          <div
            className="title"
          >
            Response
          </div>
          <div
            className="description"
          >
            Set the default response
          </div>
        </div>
      </div>
      <div
        className="active step"
        onClick={[Function]}
      >
        <div
          className="content"
        >
          <div
            className="title"
          >
            You're done
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="ui icon positive message"
      >
        <i
          aria-hidden="true"
          className="checkmark icon"
        />
        <div
          className="content"
        >
          <div
            className="header"
          >
            Congratulations !
          </div>
          <p>
            You are now ready to receive your first webhook.
          </p>
        </div>
      </div>
      <div
        className="field"
      >
        <button
          className="ui button"
          onClick={[Function]}
          wizardIdentifier="next-done-button"
        >
          Done
        </button>
      </div>
    </div>
  </div>
</div>
`;
